### PR TITLE
net processing: Move peer_map to PeerManager

### PR DESCRIPTION
**Pull request by jnewbery**
This moves `g_peer_map` from a global in net_processing.cpp's unnamed namespace to being a member `m_peer_map` of `PeerManager`.

### CODE REVIEW

**Comment by jnewbery** [comment]
Requested by @MarcoFalke @sdaftuar and @theuni . Review very much appreciated :pray:

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -156,7 +156,10 @@ static UniValue getpeerinfo(const JSONRPCRequest& request)
     for (const CNodeStats& stats : vstats) {
         UniValue obj(UniValue::VOBJ);
         CNodeStateStats statestats;
-        bool fStateStats = GetNodeStateStats(stats.nodeid, statestats);
+        bool fStateStats{false};
+        if (node.peerman) {
```
Comment (MarcoFalke) [inline src/rpc/net.cpp:160]: It is currently not possible to not have `node.peerman`. I think an `Ensure..()` like `EnsureMempool` would be good. It doesn't really make sense to call `getpeerinfo` without a peerman anyway?

File: src/net_processing.h [hunk 2]
```diff
@@ -93,7 +126,14 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
      */
     void Misbehaving(const NodeId pnode, const int howmuch, const std::string& message);
 
+    /** Get statistics from node state */
+    bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
```
Comment (MarcoFalke) [inline src/net_processing.h:130]:
```suggestion
    bool GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats);
```

nit for new code

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: src/rpc/net.cpp (hunk 1 above)
Comment (jnewbery) [reply src/rpc/net.cpp]: I've done something slightly different and checked existence at the top of this function. Let me know what you think.

File: src/net_processing.h (hunk 2 above)
Comment (jnewbery) [reply src/net_processing.h:130]: fixed. Thanks!

**Comment by hebasto** [comment]
Concept ACK.

**Comment by practicalswift** [comment]
Concept ACK: decoupling is good!

File: src/net_processing.h [hunk 3]
```diff
@@ -146,6 +176,16 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
     CTxMemPool& m_mempool;
 
     int64_t m_stale_tip_check_time; //!< Next time to check for stale tip
+
+    /** Protects m_peer_map */
+    Mutex m_peer_mutex;
+    /**
+     * Map of all Peer objects, keyed by peer id. This map is protected
+     * by the global m_peer_mutex. Once a shared pointer reference is
```
Comment (hebasto) [inline src/net_processing.h:184]: 322e87b6039877ed3457baea93cc87d78f3f4d96
```suggestion
     * by the m_peer_mutex. Once a shared pointer reference is
```

File: src/net_processing.h [hunk 4]
```diff
@@ -146,6 +176,16 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
     CTxMemPool& m_mempool;
 
     int64_t m_stale_tip_check_time; //!< Next time to check for stale tip
+
+    /** Protects m_peer_map */
```
Comment (hebasto) [inline src/net_processing.h:180]: 322e87b6039877ed3457baea93cc87d78f3f4d96
nit: This comment duplicates the code :) Consider to drop it.

**Review by hebasto — CHANGES_REQUESTED** [review]
ACK 322e87b6039877ed3457baea93cc87d78f3f4d96, I have reviewed the code and it looks OK.

File: src/net_processing.h (hunk 3 above)
Comment (jnewbery) [reply src/net_processing.h:184]: good catch. Fixed.

File: src/net_processing.h (hunk 4 above)
Comment (jnewbery) [reply src/net_processing.h:180]: Normally I'd agree with removing unnecessary commenting, but I think it's useful for every member variable to have a doxygen comment. Happy to go either way if others think it should be dropped.

**Review by hebasto — APPROVED** [review]
re-ACK d350feb2013e8db8a7464ec40670ef609d77e778

File: src/net_processing.h (hunk 4 above)
Comment (laanwj) [reply src/net_processing.h:180]: I'm for keeping it, it's better for consistency to have a comment for every one.

File: src/interfaces/node.cpp [hunk 5]
```diff
@@ -105,11 +105,13 @@ class NodeImpl : public Node
...
 
             // Try to retrieve the CNodeStateStats for each node.
-            TRY_LOCK(::cs_main, lockMain);
-            if (lockMain) {
-                for (auto& node_stats : stats) {
-                    std::get<1>(node_stats) =
-                        GetNodeStateStats(std::get<0>(node_stats).nodeid, std::get<2>(node_stats));
+            if (m_context->peerman) {
+                TRY_LOCK(::cs_main, lockMain);
+                if (lockMain) {
```
Comment (ariard) [inline src/interfaces/node.cpp:110]: Does it change something if  this check is moved above at the `m_context->connman` level ? You may save one more control flow.

File: src/net_processing.cpp [hunk 6]
```diff
@@ -953,7 +953,7 @@ void PeerManager::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
     LogPrint(BCLog::NET, "Cleared nodestate for peer=%d\n", nodeid);
 }
 
-bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats) {
+bool PeerManager::GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats) {
```
Comment (ariard) [inline src/net_processing.cpp:919]: It's a bit confusing to have `GetNodeStateStats` which is actually a member of `PeerManager`, and still have a `CNodeState`, even if inside this function you effectively fetch it. Maybe rename `GetNodeProcessingStats` ?

**Review by ariard — COMMENTED** [review]
Concept ACK

File: src/interfaces/node.cpp (hunk 5 above)
Comment (hebasto) [reply src/interfaces/node.cpp:110]: > Does it change something if this check is moved above at the `m_context->connman` level ? You may save one more control flow.

Why increase code block with locked `::cs_main`?

File: src/net_processing.cpp (hunk 6 above)
Comment (jnewbery) [reply src/net_processing.cpp:919]: I think the name right now is fine. It's getting stats related to the node's state.

File: src/interfaces/node.cpp (hunk 5 above)
Comment (jnewbery) [reply src/interfaces/node.cpp:110]: > Does it change something if this check is moved above at the m_context->connman level ? You may save one more control flow.

Yes, it's a try lock. It's possible that we're not able to take the lock, and therefore that we're not able to get the node state stats, but we'd still get the node stats. If we move the try above the m_context->connman level, then if we couldn't get the cs_main lock, we wouldn't be able to get any stats.

Eventually, none of the statistics should require cs_main, so the try lock can be removed entirely.

File: src/net_processing.cpp [hunk 7]
```diff
@@ -915,8 +871,8 @@ void PeerManager::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
         PeerRef peer = GetPeerRef(nodeid);
         assert(peer != nullptr);
         misbehavior = WITH_LOCK(peer->m_misbehavior_mutex, return peer->m_misbehavior_score);
-        LOCK(g_peer_mutex);
-        g_peer_map.erase(nodeid);
+        LOCK(m_peer_mutex);
```
Comment (theuni) [inline src/net_processing.cpp:874]: Using ```GetPeerRef()``` with an encapsulated lock for ```m_peer_map```, then manually locking the private mutex 3 lines later looks strange. It also requires two locks... all the drawbacks of mutex encapsulation without exploiting the benefits :)

Since misbehavior changes to the peer will go unobserved after this point anyway, why not just take and remove with a single lock?

Something like (untested):
```c++
PeerRef PeerManager::RemovePeer(NodeId id)
{
    PeerRef ret;
    {
        LOCK(m_peer_mutex);
        auto it = m_peer_map.find(id);
        if (it != m_peer_map.end()) {
            ret = std::move(it->second);
            m_peer_map.erase(it);
        }
    }
    return ret;
}
```

Comment (jnewbery) [reply src/net_processing.cpp]: Yes, that's much better. Thank you!

I've implemented this within `GetPeerRef()` to avoid the code duplication, and also added another commit that tidies up `FinalizeNode()` a bit more. Let me know what you think.

Comment (jnewbery) [reply src/net_processing.cpp]: oops. There's a bug in the code above which caused this to fail. You can't move the shared_ptr out of the map and then erase it. More precisely, I think something like the following is happening:

- `ret = std::move(it->second);` means `ret` is reset to be a shared pointer to `Peer`. It steals the refcount from `it->second`, since `std::move(it->second)` is an rvalue. `std::move` doesn't do anything to the map entry.
- the entry is then erased from the map, decrementing the refcount to zero and destructing the `Peer` object.

Removing the `std::move` fixed this.

File: src/net_processing.cpp [hunk 8]
```diff
@@ -864,23 +864,18 @@ void PeerManager::ReattemptInitialBroadcast(CScheduler& scheduler) const
 }
 
 void PeerManager::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
-    fUpdateConnectionTime = false;
+    PeerRef peer = RemovePeer(nodeid);
```
Comment (theuni) [inline src/net_processing.cpp:867]: This undoes this assumption: https://github.com/bitcoin/bitcoin/pull/19607#discussion_r467079107 , which I'm still not convinced is a good idea.

Comment (jnewbery) [reply src/net_processing.cpp:867]: ok, removed that commit

File: src/net_processing.cpp (hunk 7 above)
Comment (theuni) [reply src/net_processing.cpp]: This is really uninteresting minutia, but @jnewbery and I happened to get into a discussion/debate about it today so I'm following up with details. Everyone else can safely ignore this as boring.

I don't believe the comment above is correct. If removing the ```std::move``` helps, I'm wondering if there might be another issue causing problems.

> - ret = std::move(it->second); means ret is reset to be a shared pointer to Peer. It steals the refcount from it->second, since std::move(it->second) is an rvalue. std::move doesn't do anything to the map entry.

Indeed ```std::move``` doesn't do anything to the map entry as it's essentially just a cast, but the ```operator=``` does likely (definitely?) reset it.
See [libc++'s implementation of shared_ptr's move assignment operator](https://github.com/llvm/llvm-project/blob/master/libcxx/include/memory#L3967), which [calls the move constructor](https://github.com/llvm/llvm-project/blob/master/libcxx/include/memory#L3845), where the rhs is null'd.


> - the entry is then erased from the map, decrementing the refcount to zero and destructing the Peer object.

We can write a quick test to see that the refcount never hits zero after removal from the map, that copy vs. move refcounts are different as we would expect them to be, and that a shared_ptr moved from a map's value is still safe to use:
```c++
#include <cassert>
#include <cstdio>
#include <map>
#include <memory>
#include <string>

static std::shared_ptr<std::string> copytest()
{
    std::map<char, std::shared_ptr<std::string> > foo;
    foo.emplace('c', std::make_shared<std::string>("hello"));
    auto it = foo.find('c');
    std::shared_ptr<std::string> ret;
    if (it != foo.end()) {
        assert(it->second.use_count() == 1);
        ret = it->second;
        assert(it->second != nullptr); // <------------------
        assert(ret.use_count() == 2);  // <------------------
        foo.erase(it);
        assert(ret.use_count() == 1);
    }
    return ret;
}

static std::shared_ptr<std::string> movetest()
{
    std::map<char, std::shared_ptr<std::string> > foo;
    foo.emplace('c', std::make_shared<std::string>("world"));
    auto it = foo.find('c');
    std::shared_ptr<std::string> ret;
    if (it != foo.end()) {
        assert(it->second.use_count() == 1);
        ret = std::move(it->second);
        assert(it->second == nullptr); // <------------------
        assert(ret.use_count() == 1);  // <------------------
        foo.erase(it);
        assert(ret.use_count() == 1);
    }
    return ret;
}

int main()
{
    auto copyptr = copytest();
    assert(copyptr.use_count() == 1);
    auto moveptr = movetest();
    assert(moveptr.use_count() == 1);

    printf("%s %s!\n", copyptr->c_str(), moveptr->c_str());
}
```

Note that this doesn't actually prove that the map itself is still in working order after the move+removal, but I don't see why that should break anything.

Lastly, we have cases in our code where we move from a map's value. One example: https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L1012

**tl;dr**: If moving a ```shared_ptr``` out of a mapped value causes a problem that a copy doesn't, we should make sure to understand why and potentially audit other places where we do it as well.

Comment (theuni) [reply src/net_processing.cpp]: Aha, I think I found the problem. From your previous version of this commit:
```c++
...
    if (it != m_peer_map.end()) {
        ret = std::move(it->second);
        if (erase) m_peer_map.erase(it);
    }
    return ret;
```
Moving only makes sense when not erasing.

File: src/net_processing.cpp [hunk 9]
```diff
@@ -909,11 +907,16 @@ void PeerManager::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
     LogPrint(BCLog::NET, "Cleared nodestate for peer=%d\n", nodeid);
 }
 
-PeerRef PeerManager::GetPeerRef(NodeId id)
+PeerRef PeerManager::GetPeerRef(NodeId id, bool erase)
```
Comment (theuni) [inline src/net_processing.cpp:910]: This is (mostly) a style preference, but I'm not a fan of adding a boolean option to substantially change the functionality of a function as opposed to just creating two functions. GetPeerRef() is logically a read-only operation.

It's also easier easy to slip up and make mistakes like this one: https://github.com/bitcoin/bitcoin/pull/19910#discussion_r499754354 ;)

File: src/net_processing.cpp (hunk 7 above)
Comment (jnewbery) [reply src/net_processing.cpp]: Indeed, the failure was caused by that. Sorry for the noise (and for casting aspersions on your code!)

File: src/net_processing.cpp (hunk 9 above)
Comment (jnewbery) [reply src/net_processing.cpp]: You got me. I've made it a separate function.

> GetPeerRef() is logically a read-only operation.

Good point. I've made it `const` (and made the mutex `mutable`)

**Comment by jnewbery** [comment]
Thanks for the review @theuni. I've addressed your comment.

**Comment by jnewbery** [comment]
Rebased

**Comment by jnewbery** [comment]
Rebased

**Review by promag — COMMENTED** [review]
Code review ACK 3bf59172859b484405321d9778da16e41dc049cc.

This is a simple and nice refactor. Verified moved code.

**Comment by jnewbery** [comment]
rebased

File: src/net_processing.h [hunk 10]
```diff
@@ -143,6 +143,10 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
      *  May return nullptr if the Peer object can't be found. */
     PeerRef GetPeerRef(NodeId id) const;
 
+    /** Get a shared pointer to the Peer object and remove it from m_peer_map.
+     *  May return nullptr if the Peer object can't be found. */
```
Comment (hebasto) [inline src/net_processing.h:147]: aedd418fed7abfd00cc731bd57ed5b013259eb5a

To be precise, it may return an [empty](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) shared pointer rather `nullptr`.

File: src/net_processing.cpp [hunk 11]
```diff
@@ -835,11 +835,13 @@ void PeerManager::FinalizeNode(const CNode& node, bool& fUpdateConnectionTime) {
...
 
-PeerRef PeerManager::GetPeerRef(NodeId id)
+PeerRef PeerManager::GetPeerRef(NodeId id) const
 {
+    PeerRef ret;
     LOCK(m_peer_mutex);
     auto it = m_peer_map.find(id);
-    return it != m_peer_map.end() ? it->second : nullptr;
+    if (it != m_peer_map.end()) ret = it->second;
+    return ret;
```
Comment (hebasto) [inline src/net_processing.cpp]: 6bfcef3797d4d105180e325033929684f84cfc08

Why are these changes required? What are benefits to return an empty shared pointer instead `nullptr`-constructed one? Btw, in all caller places the return value is checked for `nullptr`, not being empty.

**Review by hebasto — APPROVED** [review]
re-ACK aedd418fed7abfd00cc731bd57ed5b013259eb5a, added two last commits since my [previous](https://github.com/bitcoin/bitcoin/pull/19910#pullrequestreview-484014033) review.

Still required rebasing due to the conflict with #20530.

**Comment by dongcarl** [comment]
Code Review ACK aedd418

-----

Private members >> Anon-namespaced statics/functions

Especially when the functionality is strongly intertwined with the class of which it's being made a member.

File: src/net_processing.h (hunk 10 above)
Comment (jnewbery) [reply src/net_processing.h:147]: fixed

File: src/net_processing.cpp (hunk 11 above)
Comment (jnewbery) [reply src/net_processing.cpp]: This makes is clear that NRVO can be used, since there's only one return statement. I'm not sure if the previous version would have used copy elision.

**Comment by jnewbery** [comment]
Rebased and addressed outstanding review comments.

File: src/net_processing.cpp [hunk 12]
```diff
@@ -790,11 +790,9 @@ void PeerManager::FinalizeNode(const CNode& node, bool& fUpdateConnectionTime) {
     LOCK(cs_main);
     int misbehavior{0};
     {
-        PeerRef peer = GetPeerRef(nodeid);
+        PeerRef peer = RemovePeer(nodeid);
```
Comment (promag) [inline src/net_processing.cpp:793]: This also avoids duplicate lookup on m_peer_map :+1:

**Review by promag — COMMENTED** [review]
Core review ACK 4fe77ae9aab73d8de6c40d504e079e30eff45f91.

Could squash last commit or fix the comment in  5a1f08ce1f4d49fe85b8571cd5da1b1096339a89.

File: src/net_processing.cpp [hunk 13]
```diff
@@ -842,11 +790,9 @@ void PeerManager::FinalizeNode(const CNode& node, bool& fUpdateConnectionTime) {
...
         assert(peer != nullptr);
```
Comment (theuni) [inline src/net_processing.cpp:794]: This is a little awkward because of the comment in RemovePeer:
```c++
/** Get a shared pointer to the Peer object and remove it from m_peer_map.
 *  May return an empty shared_ptr if the Peer object can't be found. */
```
Might be worth of a comment explaining that _at this point_ the peer is guaranteed to exist, but not afterwards. Then again, that's kinda what the assert means. Shrug.

**Review by theuni — APPROVED** [review]
ACK 4fe77ae9aab73d8de6c40d504e079e30eff45f91 or squashed as @promag suggested.

File: src/net_processing.cpp (hunk 11 above)
Comment (theuni) [reply src/net_processing.cpp]: +1 for considering this!

I'd be curious to know if compilers are allowed to rearrange the code before deciding how many return statements there are. I could see this going either way.

**Review by hebasto — APPROVED** [review]
re-ACK 4fe77ae9aab73d8de6c40d504e079e30eff45f91, only rebased since my [previous](https://github.com/bitcoin/bitcoin/pull/19910#pullrequestreview-543818354) review and added a commit to fix comments.

File: src/net_processing.cpp (hunk 11 above)
Comment (hebasto) [reply src/net_processing.cpp]: Considering the before-change code, to be precise, in C++17 in a return statement, when the operand is a prvalue (the result of the ternary conditional expression), elision of copy/move operations is [mandatory](https://en.cppreference.com/w/cpp/language/copy_elision).

Comment (jnewbery) [reply src/net_processing.cpp]: Thanks @hebasto! Yes, I think you're right. If I'm reading it correctly, the result of the conditional operator is a prvalue (item 5 in https://en.cppreference.com/w/cpp/language/operator_other#Conditional_operator since 4 doesn't apply), and copy elision is mandatory for returning a prvalue.

I've reverted this.

Comment (hebasto) [reply src/net_processing.cpp]: > ... and copy elision is mandatory for returning a prvalue.

Good to have C++17 :)

File: src/net_processing.cpp (hunk 13 above)
Comment (jnewbery) [reply src/net_processing.cpp:794]: Yeah, I think it's ok for now. Happy to change this if you decide you feel more strongly about it.

**Comment by jnewbery** [comment]
Thanks for the review @hebasto @promag @theuni and @dongcarl. I've addressed all comments, so this should now be ready for re-review.

File: src/net_processing.cpp [hunk 14]
```diff
@@ -887,7 +833,26 @@ void PeerManager::FinalizeNode(const CNode& node, bool& fUpdateConnectionTime) {
...
+{
+    PeerRef ret;
+    LOCK(m_peer_mutex);
+    auto it = m_peer_map.find(id);
+    if (it != m_peer_map.end()) {
+        ret = std::move(it->second);
+        m_peer_map.erase(it);
+    }
+    return ret;
+}
```
Comment (hebasto) [inline src/net_processing.cpp:853]: 3025ca9e7743d9b96c22e9c6ed7ef051dcea7e54
I'm curios is it possible to force the mandatory copy/move elision for the return value here as well?

**Review by hebasto — APPROVED** [review]
re-ACK 3025ca9e7743d9b96c22e9c6ed7ef051dcea7e54, since my [previous](https://github.com/bitcoin/bitcoin/pull/19910#pullrequestreview-545574237) review only reverted the change that introduced NRVO in `PeerManager::GetPeerRef`, and comments are fixed in the proper commits.

**Comment by theuni** [comment]
Re-ACK 3025ca9e7743d9b96c22e9c6ed7ef051dcea7e54.

**Comment by dongcarl** [comment]
Re-ACK 3025ca9

-----

1. Disambiguated between nullptr and empty shared_ptr
2. Reverted refactoring in PeerManager::GetPeerRef
3. Made `Peer` constructor `explicit`

File: src/net_processing.cpp (hunk 13 above)
Comment (rebroad) [inline src/net_processing.cpp:794]: I'm getting bitcoind crashing due to this assert. It appears that sometimes FinalizeNode gets called twice (from DeleteNode()). Perhaps nRefCount is being increased between the two occurances. This only seems to happen though when there's a delay between them due to UpdateTip(). It's probably a bug I've introduced somewhere, but just leaving this comment as a "heads up".

Comment (jnewbery) [reply src/net_processing.cpp:794]: Thanks. I expect if the bug was present in master we'd have seen it in CI or from a user report.
